services/director-v2/src/simcore_service_director_v2/modules/redis.py

```diff
@@ -0,0 +1,180 @@
+import asyncio
+import logging
+from asyncio import Task
+from contextlib import asynccontextmanager, suppress
+from dataclasses import dataclass
+from functools import cached_property
+from typing import AsyncIterator, Final, Optional
+
+from fastapi import FastAPI
+from pydantic import NonNegativeInt
+from redis.asyncio import Redis
+from redis.asyncio.lock import Lock
+from settings_library.redis import RedisSettings
+from tenacity import retry
+from tenacity.before_sleep import before_sleep_log
+from tenacity.wait import wait_random
+
+from ..core.errors import ConfigurationError
+
+DEFAULT_LOCKS_PER_NODE: Final[int] = 2
+EXTEND_TASK_ATTR_NAME = "extend_task"
+
+DockerNodeId = str
+
+
+logger = logging.getLogger(__name__)
+
+redis_retry_policy = dict(
+    wait=wait_random(5, 10),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
+    reraise=True,
+)
+
+
+def setup(app: FastAPI):
+    @retry(**redis_retry_policy)
+    async def on_startup() -> None:
+        app.state.redis_lock_manager = await RedisLockManager.create(app)
+
+    async def on_shutdown() -> None:
+        lock_manager: RedisLockManager = app.state.redis_lock_manager
+        await lock_manager.close()
+
+    app.add_event_handler("startup", on_startup)
+    app.add_event_handler("shutdown", on_shutdown)
+
+
+@dataclass
+class LocksPerNodeProvider:
+    # NOTE: added here for future use, in this iteration just
+    # returns a constant
+
+    async def get(  # pylint: disable=unused-argument,no-self-use
+        self, docker_node_id: DockerNodeId
+    ) -> int:
+        return DEFAULT_LOCKS_PER_NODE
+
+
+@dataclass
+class RedisLockManager:
+    app: FastAPI
+    redis: Redis
+    lock_per_node_provider: LocksPerNodeProvider
+    lock_timeout: float = 10.0
+
+    @classmethod
+    async def create(cls, app: FastAPI) -> "RedisLockManager":
+        settings: RedisSettings = app.state.settings.REDIS
+        redis = Redis.from_url(settings.dsn_locks)
+        lock_per_node_provider = LocksPerNodeProvider()
+        return cls(app=app, redis=redis, lock_per_node_provider=lock_per_node_provider)
+
+    @classmethod
+    def instance(cls, app: FastAPI) -> "RedisLockManager":
+        if not hasattr(app.state, "redis_lock_manager"):
+            raise ConfigurationError(
+                "RedisLockManager client is not available. Please check the configuration."
+            )
+        return app.state.redis_lock_manager
+
+    @staticmethod
+    def _get_node_slots_key(docker_node_id: DockerNodeId) -> str:
+        return f"lock_manager.{docker_node_id}.lock_slots"
```

Comment:
MINOR: for constants in identifiers I tend to use class names. I find it easier to debug when I look at redis(or any other tool) database.

Moreover, you do not have any other element than ``model_slots`` so I would simplyf the name

```python
   @classmethod
    def _get_key(cls, docker_node_id: DockerNodeId) -> str:
        return f"{cls.__name__}.{docker_node_id}.lock_slots"
    
    @classmethod
    def _get_lock_name(
        cls, docker_node_id: DockerNodeId, slot_index: NonNegativeInt
    ) -> str:
        return f"{cls._get_key(docker_node_id)}.{slot_index}"
```

Reply:
This looks better, thanks for the tip!